File: Python/url_capture.py
import sys
from bs4 import BeautifulSoup
import requests


"""
The limitation of the programs are
- it will not capture any urls present outside the 'a' tags, for example paragraph with manually return url
- it will not capture the urls that doesn't contain an href with https:// or http:
"""

'''
function that checks the valid url
'''
def valid_url(string):
    # checking if the string contains the https:// or http://
    if 'http://' in string or 'https://' in string:
        return True
    else:
        return False

'''
Function to capture the external URLS and return the total number of external urls captured
'''
def capture_external_url(html_page):
    # list to store the external urls
    external_url_list = []
    # find all the a tags with href
    all_urls = html_page.find_all("a", href=True)
    for i in all_urls:
        if valid_url(i['href']):
            external_url_list.append(i)
    return len(external_url_list)

# capturing the second urls
# capturing more urs

'''
Capture the urls from the system arguments, with the help of beautifulsoup
'''
def run_scrapper(n):
    for i in range(1, n):
        # capture the url and return the results
        anticipated_url = sys.argv[i]
        # verify whether its a url or not
        if valid_url(anticipated_url):
            # capture the html page using bs4
            page = requests.get(anticipated_url)
            soup = BeautifulSoup(page.content, "html.parser")
            url_length = capture_external_url(soup)
            # print out to console
            print(str(sys.argv[i])+ " " + str(url_length))
        else:
            # else return invalid url
            print("URL is not a valid url", anticipated_url)


if __name__ == "__main__":
    # get the total number of url link present
    n = len(sys.argv)
    run_scrapper(n)

File: Javascript/url_capture.js
const puppeteer = require('puppeteer');

/*
The limitation of the programs are
- It requires a headless browser, if the installed browser is different than the one used by puppeteer, it will not work
- it will capture all the internal URL links, as it is capturing the urls from browser
*/

// check if the given url is a vaid url or not
function validURL(url) {
  if (url.includes('http://') || url.includes('https://')) {
    return true;
  } else {
    return false;
  }
}

async function run() {
  // going through all the provided arguments from the user, ignore first 2 because those are not an url
  for (let i = 2; i < process.argv.length; ++i) {
    // check if the given value is valid url or not
    current_url = process.argv[i];
    const validResult = validURL(current_url);

    //  if the given url is a vaid url capture the results
    if (validResult === true) {
      // storing the href
      var hrefArray = new Array();

      // capture the html page with other parser around, do nothing about it
      const browser = await puppeteer.launch();
      const page = await browser.newPage();
      await page.goto(current_url);

      const hrefs = await page.$$eval('a', (as) => as.map((a) => a.href));
      console.log(hrefs);

      for (let urls of hrefs) {
        if (urls.includes('http://') || urls.includes('https://')) {
          hrefArray.push(urls);
        }
      }

      // capture the html page
      await browser.close();

      // log this to console
      console.log(current_url + ' ' + hrefArray.length);
    } else {
      // else throw the invalid url to the user
      console.log('Invalid URL', current_url);
    }
  }
}

run();
